test/backend_tests.cc

```diff
@@ -0,0 +1,404 @@
+	const std::vector<uint8_t> expected(size, sycl_backend_detail::uninitialized_memory_pattern);
+
+	for(const auto did : std::initializer_list<std::optional<device_id>>{std::nullopt, device_id(0)}) {
+		CAPTURE(did);
+		const auto ptr = backend_alloc(*backend, did, 1024, 1);
+		std::vector<uint8_t> contents(size);
+		sycl_queue.memcpy(contents.data(), ptr, size).wait();
+		CHECK(contents == expected);
+		backend_free(*backend, did, ptr);
+	}
+#else
+	SKIP("Not in a debug build");
+#endif
+}
+
+TEST_CASE_METHOD(test_utils::backend_fixture, "host task lambdas are hydrated and invoked with the correct parameters", "[backend]") {
+	const auto [backend_type, backend, sycl_devices] = generate_backends_with_devices();
+	CAPTURE(backend_type, sycl_devices);
+
+	const mock_accessor<target::host_task> acc1(hydration_id(1));
+	const mock_accessor<target::host_task> acc2(hydration_id(2));
+	const std::vector<closure_hydrator::accessor_info> accessor_infos{
+	    {reinterpret_cast<void*>(0x1337), box<3>{{1, 2, 3}, {4, 5, 6}},
+	        box<3>{{0, 1, 2}, {7, 8, 9}} CELERITY_DETAIL_IF_ACCESSOR_BOUNDARY_CHECK(, reinterpret_cast<oob_bounding_box*>(0x69420))},
+	    {reinterpret_cast<void*>(0x7331), box<3>{{3, 2, 1}, {6, 5, 4}},
+	        box<3>{{2, 1, 0}, {9, 8, 7}} CELERITY_DETAIL_IF_ACCESSOR_BOUNDARY_CHECK(, reinterpret_cast<oob_bounding_box*>(0x1230))}};
+
+	constexpr size_t lane = 0;
+	const box<3> execution_range({1, 2, 3}, {4, 5, 6});
+	const auto collective_comm = reinterpret_cast<const communicator*>(0x42000);
+
+	int value = 1;
+
+	// no accessors
+	test_utils::await(backend->enqueue_host_task(
+	    lane,
+	    [&](const box<3>& b, const communicator* c) {
+		    CHECK(b == execution_range);
+		    CHECK(c == collective_comm);
+		    value += 1;
+	    },
+	    {}, execution_range, collective_comm));
+
+	// yes accessors
+	test_utils::await(backend->enqueue_host_task(
+	    lane,
+	    [&, acc1, acc2](const box<3>& b, const communicator* c) {
+		    REQUIRE(acc1.info.has_value());
+		    REQUIRE(acc2.info.has_value());
+		    CHECK(accessor_info_equal(*acc1.info, accessor_infos[0]));
+		    CHECK(accessor_info_equal(*acc2.info, accessor_infos[1]));
+		    CHECK(b == execution_range);
+		    CHECK(c == collective_comm);
+		    value += 1;
+	    },
+	    accessor_infos, execution_range, collective_comm));
+
+	CHECK(value == 3);
+}
+
+TEST_CASE_METHOD(test_utils::backend_fixture, "host tasks in a single lane execute in-order", "[backend]") {
+	const auto [backend_type, backend, sycl_devices] = generate_backends_with_devices();
+	CAPTURE(backend_type, sycl_devices);
+
+	constexpr size_t lane = 0;
+	const auto first = backend->enqueue_host_task(
+	    lane, [](const box<3>&, const communicator*) { std::this_thread::sleep_for(std::chrono::milliseconds(10)); }, {}, box_cast<3>(box<0>()), nullptr);
+	const auto second = backend->enqueue_host_task(
+	    lane, [](const box<3>&, const communicator* /* collective_comm */) {}, {}, box_cast<3>(box<0>()), nullptr);
```

Comment:
I like it!

Reply:
What I meant was to already check inside `second` whether the ID was set by the first task, because only then you can be sure about the ordering (whereas the loop is technically racy). But again, I don't think it's that important 😅

Reply:
Added another check for this.